src/rust_connection/fd_read_write.rs

```diff
@@ -379,8 +349,15 @@ impl<R: ReadFD> ReadFD for BufReadFD<R> {
         self.start += nread;
         Ok(nread)
     }
+}
 
-    fn set_nonblocking(&mut self, nonblocking: bool) -> Result<()> {
-        self.inner.set_nonblocking(nonblocking)
+impl<T: ReadFD> Poll for BufReadFD<T> {
+    fn poll(&mut self, read: bool, write: bool) -> Result<(bool, bool)> {
+        if read && self.start < self.end {
+            // Avoid blocking poll if buffers are not empty.
```

Comment:
I love that this just magically falls into place with this new trait. With my ideas, the "some unread data is still buffered"-case was always a problem.

:+1: